### ISSUE TITLE
`systemd-nspawn` fails to resolve altname with `--network-interface=` on older kernels

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#31104]
This seems to be fallout from https://github.com/systemd/systemd/pull/30956, but CI didn't catch it because Ubuntu CI was gone and the C8S job runs QEMU tests only once per day (where it failed, but I got to it just now). It can be easily reproduced with:

```
# ip link add dev sd-test type dummy && ip link property add dev sd-test altname foo; ip link show sd-test
6: sd-test: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether be:0e:8b:21:c8:ca brd ff:ff:ff:ff:ff:ff
    altname foo
# systemd-nspawn -D / --volatile=yes --network-interface=sd-test:foo true
░ Spawning container centos8s.localdomain on /.
░ Press Ctrl-] three times within 1s to kill container.

Unknown interface name foo: No such device
(sd-netns) failed with exit status 1.
Failed to fork process (sd-netns): Protocol error
```

It's the `rtnl_resolve_link_alternative_name()` call from `sd_device_new_from_ifname()` that fails in this case, but that's as far as I've gotten. /cc @yuwata

**Comment by yuwata** [comment]
Hm? Could you provide a link to the failed log?

**Comment by yuwata** [comment]
Does the kernel include the following commit?
https://github.com/torvalds/linux/commit/76c9ac0ee878f6693d398d3a95ccaf85e1f597a

**Comment by yuwata** [comment]
Below is in v5.19,
https://github.com/torvalds/linux/commit/5ea08b5286f66ee5ac0150668c92d1718e83e1ad (v5.19)
But, if I believe the commit message, this should not change any behavior for RTM_GETLINK.

**Comment by yuwata** [comment]
@jpirko Do you have any idea about that?
The issue is in an old kernel, that we cannot resolve an interface from an alternative interface name (may be in a 'non-main' network namespace, as we have test cases about that in a 'main' network namespace and they works fine if I believe the log).

**Comment by yuwata** [comment]
@mrc0mmand Does the following work on the old kernel environment?
```sh
$ sudo ip netns add hoge
$ sudo ip netns exec hoge ip link add foo type dummy
$ sudo ip netns exec hoge ip link property add dev foo altname foo-altname
$ sudo ip netns exec hoge ip link show dev foo-altname
5: foo: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 06:60:ef:a5:b9:69 brd ff:ff:ff:ff:ff:ff
    altname foo-altname
```

**Comment by yuwata** [comment]
If I understand the commit message correctly, the following commit should fix the issue:
https://github.com/torvalds/linux/commit/8e15aee621618a3ee3abecaf1fd8c1428098b7ef (v6.6).

**Comment by yuwata** [comment]
So, remaining thing here is, we should workaround the issue in userspace or not...

**Comment by yuwata** [comment]
The fix is already backported to v6.1-stable branch:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/net/core/dev.c?h=linux-6.1.y&id=46de539282f278cae142d38ae32aed3e20f07a09

But, the fix is on https://github.com/torvalds/linux/commit/5ea08b5286f66ee5ac0150668c92d1718e83e1ad (v5.19), so the currently the fix has not come to v5.15-stable branch yet.

**Comment by yuwata** [comment]
So, the fix in the kernel side is quite new. I opened PR #31109 for the issue. I have not tested it, as my laptop runs kernel 6.6. @mrc0mmand Could you test the PR?
